tests/configtest1-result.txt
configuration ok.


tests/configtest2-result.txt
tests/cpmrc-broken:5: Unknown Config-Option: 'IllegalOption'
error in configuration file.


docs/ChangeLog
Wed Jul 12 08:41:57 CEST 2006  harry_b@mm.st
  * Added new searchpatterns to list hosts, services and users and modified the password alphabet to create more keyboard friendly passwords.

Mon Jul 10 14:34:08 CEST 2006  harry_b@mm.st
  * Typo fix.

Mon Jul 10 14:33:47 CEST 2006  harry_b@mm.st
  * Fixed wrong XML code in the example.

Thu Jul  6 11:23:38 CEST 2006  harry_b@mm.st
  * Todos updated.

Wed Apr 12 10:55:37 CEST 2006  harry_b@mm.st
  * The statusline now reflects the correct ESC command which now switches from 'Exit' to 'Back' on deeper levels.

Fri Apr  7 13:42:44 CEST 2006  harry_b@mm.st
  * Added support for the current version of CDK 5.

Mon Apr  3 20:17:24 CEST 2006  harry_b@mm.st
  * Added signing of the Debian repositories.

Sun Mar 26 17:00:29 CEST 2006  harry_b@mm.st
  tagged VERSION_0_22BETA

Sun Mar 26 16:55:40 CEST 2006  harry_b@mm.st
  * Translation added.

Sun Mar 26 16:54:56 CEST 2006  harry_b@mm.st
  * Added feature to ask the user wether the lockfile should be removed or not.

Sun Mar 26 16:54:37 CEST 2006  harry_b@mm.st
  * Updated the changelog.

Sun Mar 26 15:57:09 CEST 2006  harry_b@mm.st
  * Code cleanup.

Sat Mar 25 12:03:41 CET 2006  harry_b@mm.st
  * Fixed for bug #1458208.

Sat Mar 25 11:28:13 CET 2006  harry_b@mm.st
  * Fix for bug #1457707.

Sat Mar 25 11:27:47 CET 2006  harry_b@mm.st
  * Fixed CDK header detection for gcc multi-files per line output.

Tue Feb 28 21:18:01 CET 2006  harry_b@mm.st
  * Replaced the sed call with a perl call so the CDK detection works even without a current version of sed.

Tue Feb 28 21:17:43 CET 2006  harry_b@mm.st
  * Version upgraded to 0.22beta.

Mon Feb 27 11:20:03 CET 2006  harry_b@mm.st
  tagged VERSION_0_21BETA

Mon Feb 27 11:17:29 CET 2006  harry_b@mm.st
  tagged VERSION_0_21BETA

Mon Feb 27 11:07:37 CET 2006  harry_b@mm.st
  * Fixed order of chown and chmod calls.

Mon Feb 27 11:01:52 CET 2006  harry_b@mm.st
  tagged VERSION_0_21BETA

Mon Feb 27 10:05:10 CET 2006  harry_b@mm.st
  * Todo list updated.

Mon Feb 27 09:49:44 CET 2006  harry_b@mm.st
  * The changelogs are no longer checked by lintian since the Debian specific changelog just contains a single line and produces unnecessary warnings.

Fri Feb 24 17:22:46 CET 2006  harry_b@mm.st
  * Added explicit setting of permissions for the debian postinst file.

Fri Feb 24 17:19:12 CET 2006  harry_b@mm.st
  * Modified build process for deb-packages so they now support the wrapper-less distribution as well.

Fri Feb 24 17:02:54 CET 2006  harry_b@mm.st
  * Documentation updated.

Fri Feb 24 16:59:14 CET 2006  harry_b@mm.st
  * Added automatic detection of GpgME version 1.1+ to enable automatic recipient detection. Moreover a short information was added to the --version display.

Thu Feb 23 18:05:37 CET 2006  harry_b@mm.st
  * Fixed bug #1437432.

Mon Feb 20 17:45:08 CET 2006  harry_b@mm.st
  * The manpage is now man1 instead of man8.

Mon Feb 20 15:45:34 CET 2006  harry_b@mm.st
  * Fixed copyright information and the address of the Free Software Foundation.

Sun Feb 19 22:31:17 CET 2006  harry_b@mm.st
  * Added label GPGME_HAS_RECIPIENT to suppres the stuff which is new to GPGME 1.1.

Sun Feb 19 22:30:51 CET 2006  harry_b@mm.st
  * Version updated to 0.21beta.

Sun Feb 19 22:30:37 CET 2006  harry_b@mm.st
  * Fixed deprecated chown call.

Sun Feb 19 22:30:14 CET 2006  harry_b@mm.st
  * Modified date entries and updated the changelog.

Sun Feb 19 21:59:59 CET 2006  harry_b@mm.st
  * Added automatic addition of keys when encrypted files are opened.

Sun Feb 19 21:58:48 CET 2006  harry_b@mm.st
  * Removed HAVE_REALLOC for correct usage on BSD systems. Otherwise memory problems occur there.

Wed Jan 11 08:30:11 CET 2006  harry_b@mm.st
  * Fixed libc6 dependency for Debian Sarge.

Wed Jan 11 08:29:40 CET 2006  harry_b@mm.st
  * Fixed autoconf to get it working on none-Linux platforms.

Thu Dec  1 10:24:17 CET 2005  harry_b@mm.st
  * Added support for different Debian distributions.

Thu Dec  1 10:06:35 CET 2005  harry_b@mm.st
  * Todo updated.

Thu Dec  1 09:53:44 CET 2005  harry_b@mm.st
  tagged VERSION_0_20BETA

Thu Dec  1 09:30:12 CET 2005  harry_b@mm.st
  * Added apt repository update for sarge and unstable.

Thu Dec  1 09:29:59 CET 2005  harry_b@mm.st
  * Documentation updated.

Thu Dec  1 09:29:14 CET 2005  harry_b@mm.st
  * Documentation added.

Thu Dec  1 09:08:37 CET 2005  harry_b@mm.st
  * Added the @ character to the general string alphabet (bug #1370314).

Thu Dec  1 09:08:23 CET 2005  harry_b@mm.st
  * Todo updated.

Sat Nov  5 09:08:37 CET 2005  harry_b@mm.st
  * Replaced the permission setting call.

Sat Nov  5 09:08:20 CET 2005  harry_b@mm.st
  * Another fix for the CDK header handling.

Sat Nov  5 08:45:37 CET 2005  harry_b@mm.st
  * Changelog updated.

Sat Nov  5 08:45:22 CET 2005  harry_b@mm.st
  * Added a version check for GPGME.

Sat Nov  5 08:44:44 CET 2005  harry_b@mm.st
  * Fixed the checking of the cdk.h file which could not be properly detected before.

Sat Nov  5 08:44:09 CET 2005  harry_b@mm.st
  * Todolist updated.

Sat Nov  5 08:14:49 CET 2005  harry_b@mm.st
  * Todo and documentation updated.

Sat Nov  5 07:53:07 CET 2005  harry_b@mm.st
  * Added support for the case that xmllint is not in the path or does not even exist.

Sat Nov  5 07:33:58 CET 2005  harry_b@mm.st
  * The prepare process now checks and sets filepermissions.

Sat Nov  5 07:26:02 CET 2005  harry_b@mm.st
  * Removed the error message check from the CLI search 3 check. Now only an existing error is checked but not it's exact message.

Sun Oct 30 19:51:24 CET 2005  harry_b@mm.st
  * Translation updated.

Sun Oct 30 19:48:51 CET 2005  harry_b@mm.st
  * Even if no key was specified, the first question always must be, if the user really wants to quit.

Sun Oct 23 20:30:07 CEST 2005  harry_b@mm.st
  * Documentation updated.

Sun Oct 23 20:07:38 CEST 2005  harry_b@mm.st
  * Translation updated.

Sun Oct 23 20:04:36 CEST 2005  harry_b@mm.st
  * Todo list updated.


options.c
/* #############################################################################
 * code for all commandline argument handling
 * #############################################################################
 * Copyright (C) 2005, 2006 Harry Brueckner
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or any later version.
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 * Contact: Harry Brueckner <harry_b@mm.st>
 *          Muenchener Strasse 12a
 *          85253 Kleinberghofen
 *          Germany
 * #############################################################################
 */

/* #############################################################################
 * includes
 */
#include "cpm.h"
#ifdef HAVE_GETOPT_H
  #include <getopt.h>
#endif
#ifdef HAVE_LIBNCURSES
  #include <ncurses.h>
#endif
#ifdef HAVE_LIBXML2
  #include <libxml/xmlversion.h>
#endif
#ifdef HAVE_LIBZ
  #include <zlib.h>
#endif
#include <gpgme.h>
#include "configuration.h"
#include "general.h"
#include "listhandler.h"
#include "memory.h"
#include "options.h"
#include "string.h"


/* #############################################################################
 *
 * Description    in case getOptions() did not provide any values, this
 *                function makes sure, all mandatory values are present
 * Author         Harry Brueckner
 * Date           2005-03-17
 * Arguments      void
 * Return         void
 */
void getDefaultOptions(void)
  {
    int                 found = 0;
    char*               home;

    /* Flawfinder: ignore */
    home = getenv("HOME");

    if (!config -> dbfilerc)
      {   /* we initialize the default db file */
        config -> dbfilerc = getFilePath(home, DEFAULT_DB_FILE);
      }

    if (!config -> rcfile)
      {   /* we initialize the default rc file */
        config -> rcfile = getFilePath(home, DEFAULT_RC_FILE);

        if (!fileExists(config -> rcfile))
          {   /* we don't have the user's rc file in ~/.cpmrc, so we try to
               * find it in the default RC path (/etc/cpmrc)
               */
            memFreeString(__FILE__, __LINE__, config -> rcfile);
            config -> rcfile = getFilePath(DEFAULT_RC_PATH_1,
                DEFAULT_ETC_RC_FILE);
          }
        else
          { found = 1; }
        if (!found && !fileExists(config -> rcfile))
          {   /* we didn't find it in /etc/cpmrc either, so we try
               * /etc/cpm/cpmrc
               */
            memFreeString(__FILE__, __LINE__, config -> rcfile);
            config -> rcfile = getFilePath(DEFAULT_RC_PATH_2,
                DEFAULT_ETC_RC_FILE);
          }
        else
          { found = 1; }
        if (!found && !fileExists(config -> rcfile))
          {   /* if the file still does not exist, we empty the configuration */
            memFreeString(__FILE__, __LINE__, config -> rcfile);
            config -> rcfile = NULL;
          }
      }
    if (!config -> encoding)
      {
        config -> encoding = memAlloc(__FILE__, __LINE__,
            strlen(DEFAULT_ENCODING) + 1);
        strStrncpy(config -> encoding, DEFAULT_ENCODING,
            strlen(DEFAULT_ENCODING) + 1);
      }
  }


/* #############################################################################
 *
 * Description    this function concats a path and a filename, calculating
 *                its length and allocating memory accordingly
 * Author         Harry Brueckner
 * Date           2005-03-17
 * Arguments      char* path      - path to use (can be NULL)
 *                char* filename  - filename to append to the path
 * Return         char* pointer to the allocated area
 */
char* getFilePath(char* path, char* filename)
  {
    char*               result;

    if (path)
      {   /* we got a path and a filename */
        result = memAlloc(__FILE__, __LINE__,
            strlen(path) + 1 + strlen(filename) + 1);
        strStrncpy(result, path, strlen(path) + 1);
        strStrncat(result, "/",  1 + 1);
        strStrncat(result, filename, strlen(filename) + 1);
      }
    else
      {   /* we just got a filename */
        result = memAlloc(__FILE__, __LINE__, strlen(filename) + 1);
        strStrncpy(result, filename, strlen(filename) + 1);
      }

    return result;
  }


/* #############################################################################
 *
 * Description    get the commandline options, the command was called with
 * Author         Harry Brueckner
 * Date           2005-03-16
 * Arguments      int     argc    - argument counter
 *                char**  argv    - commandline arguments
 * Return         0 if everything is ok, otherwise 1
 */
int getOptions(int argc, char **argv)
  {
    int                 error = 0;

    while (1)
      {
        int             code;
        int             optid = 0;

        /* remember to fix the wrapper script as well if those options are
         * modified.
         */
        static struct option long_options[] =
          {
            { "config",       required_argument,  0, 0 },   /*  0 */
            { "configtest",   no_argument,        0, 0 },   /*  1 */
            { "encoding",     required_argument,  0, 0 },   /*  2 */
            { "file",         required_argument,  0, 0 },   /*  3 */
            { "help",         no_argument,        0, 0 },   /*  4 */
            { "ignore",       no_argument,        0, 0 },   /*  5 */
            { "key",          required_argument,  0, 0 },   /*  6 */
            { "noencryption", no_argument,        0, 0 },   /*  7 */
            { "noignore",     no_argument,        0, 0 },   /*  8 */
            { "readonly",     no_argument,        0, 0 },   /*  9 */
            { "regex",        no_argument,        0, 0 },   /* 10 */
            { "regular",      no_argument,        0, 0 },   /* 11 */
            { "security",     no_argument,        0, 0 },   /* 12 */
            { "testrun",      required_argument,  0, 0 },   /* 13 */
            { "version",      no_argument,        0, 0 },   /* 14 */
            { 0,              0,                  0, 0 }
          };

        /* Flawfinder: ignore */
        code = getopt_long(argc, argv, "c:e:f:hirs", long_options, &optid);
        if (code == -1)
        { break; }

        if (code == 0)
          {   /* this is a long option */
            switch (optid)
              {
                case 0:   /* config */
                    code = 'c';
                    break;
                case 1:   /* configtest */
                    config -> configtest = 1;
                    break;
                case 2:   /* encoding */
                    code = 'e';
                    break;
                case 3:   /* file */
                    code = 'f';
                    break;
                case 4:   /* help */
                    code = 'h';
                    break;
                case 5:   /* ignore */
                    code = 'i';
                    break;
                case 6:   /* key */
                    if (!runtime -> commandlinekeys)
                      {   /* if we find the first key on the commandline, we
                           * free the list and start collecting those keys
                           */
                        listFree(config -> defaultkeys);
                        runtime -> commandlinekeys = 1;
                      }
                    config -> defaultkeys = listAdd(config -> defaultkeys,
                        optarg);
                    break;
                case 7:   /* noencryption */
                    config -> encryptdata = 0;
                    break;
                case 8:   /* noignore */
                    runtime -> casesensitive = 1;
                    break;
                case 9:   /* readonly */
                    config -> readonly = 1;
                    break;
                case 10:   /* regex */
                    code = 'r';
                    break;
                case 11:   /* regular */
                    runtime -> searchtype = SEARCH_REGULAR;
                    break;
                case 12:   /* security */
                    code = 's';
                    break;
                case 14:   /* version */
                    config -> version = 1;
                    break;
                case 13:   /* testrun */
#ifdef TEST_OPTION
                    if (!optarg)
                      {
                        fprintf(stderr,
                            _("error: no argument given for --testrun.\n"));
                        error = 1;
                      }
                    else
                      {
                        if (strlen(optarg) > STDSTRINGLENGTH)
                          {
                            fprintf(stderr,
                                _("error: --testrun argument too long.\n"));
                            error = 1;
                          }
                        else
                          {
                            config -> testrun = memAlloc(__FILE__, __LINE__,
                                strlen(optarg) + 1);
                            strStrncpy(config -> testrun, optarg,
                                strlen(optarg) + 1);
                          }
                      }
#else
                    fprintf(stderr,
                        _("error: the program was not compiled with the -DTEST_OPTION option.\n"));
                    error = 1;
#endif
                    break;
                default:
                    printf(
                        _("error: getopt returned unknown long option %d.\n"),
                        optid);
              }
          }

        switch (code)
          {
            case 0:   /* long argument only */
                break;
            case 'c':
                if (!optarg)
                  {
                    fprintf(stderr,
                        _("error: no filename given for --config.\n"));
                    error = 1;
                  }
                else if (strlen(optarg) > STDSTRINGLENGTH)
                  {
                    fprintf(stderr,
                        _("error: --config argument too long.\n"));
                    error = 1;
                  }
                else
                  {
                    config -> rcfile = memAlloc(__FILE__, __LINE__,
                        strlen(optarg) + 1);
                    strStrncpy(config -> rcfile, optarg, strlen(optarg) + 1);
                  }
                break;
            case 'e':
                if (!optarg)
                  {
                    fprintf(stderr,
                        _("error: no argument given for --encoding.\n"));
                    error = 1;
                  }
                else if (strlen(optarg) > STDSTRINGLENGTH)
                  {
                    fprintf(stderr,
                        _("error: --encoding argument too long.\n"));
                    error = 1;
                  }
                else
                  {
                    config -> encoding = memAlloc(__FILE__, __LINE__,
                        strlen(optarg) + 1);
                    strStrncpy(config -> encoding, optarg, strlen(optarg) + 1);
                  }
                break;
            case 'f':
                if (!optarg)
                  {
                    fprintf(stderr,
                        _("error: no filename given for --file.\n"));
                    error = 1;
                  }
                else if (strlen(optarg) > STDSTRINGLENGTH)
                  {
                    fprintf(stderr,
                        _("error: --file argument too long.\n"));
                    error = 1;
                  }
                else
                  {
                    config -> dbfilecmd = memAlloc(__FILE__, __LINE__,
                        strlen(optarg) + 1);
                    strStrncpy(config -> dbfilecmd, optarg, strlen(optarg) + 1);
                  }
                break;
            case 'h':
                config -> help = 1;
                break;
            case 'i':
                runtime -> casesensitive = 0;
                break;
            case 'r':
                runtime -> searchtype = SEARCH_REGEX;
                break;
            case 's':
                config -> security = 1;
                break;
            case ':':
            case '?':
                error = 1;
                break;
            default:
                fprintf(stderr,
                    _("error: getopt returned character code 0x%x\n"),
                    code);
                error = 1;
          }
      }

    if (optind < argc)
      {
        int             entries,
                        i;

        entries = argc - optind;
        for (i = 0; i < entries; i++, optind++)
          {
            if (strlen(argv[optind]) > STDSTRINGLENGTH)
              {
                fprintf(stderr, _("error: argument too long.\n"));
                error = 1;
              }
            else
              {
                config -> searchdata = listAdd(config -> searchdata,
                    argv[optind]);
              }
          }
      }

    /* find out wether we run in CLI or GUI mode */
    if (config -> searchdata ||
#ifdef TEST_OPTION
        config -> testrun)
#else
        0)
#endif
      { runtime -> guimode = 0; }
    else
      { runtime -> guimode = 1; }

    return error;
  }


/* #############################################################################
 *
 * Description    show the syntax help for the command
 * Author         Harry Brueckner
 * Date           2005-03-16
 * Arguments      void
 * Return         void
 */
void showHelp(void)
  {
    printf(_("usage: cpm [--config FILE] [--help] [PATH] ...\n"));
    printf(_("    --config, -c    configuration file to use [~/%s]\n"),
        DEFAULT_RC_FILE);
    printf(_("    --configtest    verify the configuration file and exit\n"));
    printf(_("    --encoding, -e  the encoding in which keyboard input arrives [%s]\n"),
        DEFAULT_ENCODING);
    printf(_("    --file, -f      database file to use [~/%s]\n"),
        DEFAULT_DB_FILE);
    printf(_("    --help, -h      display this help\n"));
    printf(_("    --ignore, -i    search case insensitive in cli mode\n"));
    printf(_("    --key           overwrite the default encryption keys and use this key\n"));
    printf(_("                    instead; repeat for several keys\n"));
    printf(_("    --noencryption  turn off file encryption\n"));
    printf(_("                    WARNING: THIS IS FOR DEVELOPMENT AND TESTING ONLY!\n"));
    printf(_("    --noignore      search case sensitive in cli mode\n"));
    printf(_("    --readonly      open the database in read-only mode\n"));
    printf(_("    --regex, -r     search with regular expressions in cli mode\n"));
    printf(_("    --regular       use regular search in cli mode\n"));
    printf(_("    --security, -s  run a security check and show the current security status\n"));
#ifdef TEST_OPTION
    printf(_("    --testrun       run one of the testmodes\n"));
    printf(_("                    backup        - run test on the backupfile creation\n"));
    printf(_("                    compress[1-6] - run compression test 1-6\n"));
    printf(_("                    decrypt       - run test on the decryption code\n"));
    printf(_("                    encrypt       - run test on the encryption code\n"));
    printf(_("                    environment   - run test on the environment validation\n"));
    printf(_("                    garbage       - run test on garbage input files\n"));
    printf(_("                    searchpattern - run test on the search patterns\n"));
#endif
    printf(_("    --version       display the version and exit\n"));
    printf(_("    PATH            path to display the password for\n"));
  }


/* #############################################################################
 *
 * Description    show the syntax help for the command
 * Author         Harry Brueckner
 * Date           2005-03-16
 * Arguments      void
 * Return         void
 */
void showVersion(void)
  {
    printf("%s\n", PACKAGE_STRING);
#ifdef CDK_VERSION_H
    printf(_("CDK version %s.%s (%s).\n"),
        CDK_VERSION_MAJOR,
        CDK_VERSION_MINOR,
        CDK_VERSION_PATCH);
#else
    printf(_("CDK version 4 (unknown version).\n"));
#endif
#ifdef GPGME_HAS_RECIPIENT
    printf(_("GpgME version %s (rcpt).\n"), GPGME_VERSION);
#else
    printf(_("GpgME version %s.\n"), GPGME_VERSION);
#endif
    printf(_("ncurses version %s (%ld).\n"),
        NCURSES_VERSION, NCURSES_VERSION_PATCH);
    printf(_("XML2 version %s.\n"), LIBXML_DOTTED_VERSION);
    printf(_("zlib version %s.\n"), ZLIB_VERSION);
#ifdef HAVE_CRACKLIB
    printf(_("cracklib is enabled.\n"));
#else
    printf(_("cracklib is disabled.\n"));
#endif
    printf(_("Written by Harry Brueckner <harry_b@mm.st>.\n"));
  }


/* #############################################################################
 */

